sdk/core/azure-mgmt-core/azure/mgmt/core/policies/_base.py

```diff
@@ -0,0 +1,119 @@
+# --------------------------------------------------------------------------
+#
+# Copyright (c) Microsoft Corporation. All rights reserved.
+#
+# The MIT License (MIT)
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the ""Software""), to
+# deal in the Software without restriction, including without limitation the
+# rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
+# sell copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED *AS IS*, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+# FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+# IN THE SOFTWARE.
+#
+# --------------------------------------------------------------------------
+import json
+import logging
+import re
+import time
+import uuid
+
+from azure.core.pipeline import PipelineContext, PipelineRequest
+from azure.core.pipeline.policies import HTTPPolicy
+from azure.core.pipeline.transport import HttpRequest
+
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class ARMAutoResourceProviderRegistrationPolicy(HTTPPolicy):
+    """Auto register an ARM resource provider if not done yet.
+    """
+
+    def send(self, request):
+        # type: (PipelineRequest[HTTPRequestType], Any) -> PipelineResponse[HTTPRequestType, HTTPResponseType]
+        http_request = request.http_request
+        response = self.next.send(request)
+        if response.http_response.status_code == 409:
+            rp_name = self._check_rp_not_registered_err(response)
+            if rp_name:
+                url_prefix = self._extract_subscription_url(http_request.url)
+                if not self._register_rp(request, url_prefix, rp_name):
+                    return response
+                # Change the 'x-ms-client-request-id' otherwise the Azure endpoint
+                # just returns the same 409 payload without looking at the actual query
+                if "x-ms-client-request-id" in http_request.headers:
```

Comment:
the RequestIdPolicy is after the ARMAutoResourceProviderRegistrationPolicy in default pipeline. I am afreaid the code may not work as expected.

Reply:
That's a good point, thinking about it I will swap them, thanks!